### ISSUE TITLE
Still Laravel 4 compatible?

### ISSUE THREAD

**Issue by MarcosBL** [elasticquent/Elasticquent#63]
I'm in the need of extending an existing L4 app, and tried to follow the setup procedure, but as soon as I copy the "providers" declaration, I get a:

```php
Call to undefined method Elasticquent\ElasticquentServiceProvider::publishes()
```

pointing to

```php
 public function boot()
    {
        $this->publishes([
            __DIR__.'/config/elasticquent.php' => config_path('elasticquent.php'),
        ]);
```

I have an **/app/config/elasticquent.php** file in place, but that doesn't seem to make any difference.

Is this a L5-only-app problem? Any specific branch I can target to get L4 support?

Thanks in advance!

**Comment by chuangbo** [comment]
I'll fix it now.

**Comment by chuangbo** [comment]
Well I was trying to make it compatible with laravel 4, but I'm not familiar with it.

So I think there is two temporary solution for you:

1. Use version 1.0.6
2. Remove `Elasticquent\ElasticquentServiceProvider` in your `config/app.php`
3. Remove facade `Elasticquent\ElasticquentElasticsearchFacade` in your `config/app.php`

**Comment by MarcosBL** [comment]
Thank you very much for the effort!

Have had no luck with the 1.0.6 version, seems to be working with 1.0.5, using that concrete readme docs

**Comment by timgws** [comment]
I will look into this in the next two days, and fix it up. I don't have any L4 apps using ES at the moment.

**Comment by MarcosBL** [comment]
Thank you very much @timgws , for now it's working flawlessly with 1.0.5

**Comment by MarcosBL** [comment]
Uops... I'll have to wait till you have the time to take a look at 1.0.6 L4 incompatibility, as I've been to fast: there is no support for indexSettings in 1.0.5 :(

**Comment by chuangbo** [comment]
@MarcosBL Try to use latest version (dev-master) but not `Elasticquent\ElasticquentServiceProvider`?

**Comment by MarcosBL** [comment]
That circumvents the publishes() problem, but shows another incompatibility, as my models extend Ardent (not sure where the problem resides, Ardent or Elasticquent):

```
<?php
use LaravelBook\Ardent\Ardent;
use Cviebrock\EloquentSluggable\SluggableInterface;
use Cviebrock\EloquentSluggable\SluggableTrait;
use Elasticquent\ElasticquentTrait;

class Textos extends Ardent implements SluggableInterface {
	use ElasticquentTrait;

```

It works fine this way in 1.0.5, but in dev-master:

```
PHP Fatal error:  Declaration of Elasticquent\ElasticquentTrait::newQuery() must be compatible with LaravelBook\Ardent\Ardent::newQuery($excludeDeleted = true)
```

I narrowed this to https://github.com/elasticquent/Elasticquent/blob/master/src/EloquentMethods.php#L33

Adding the required parameter, 

```abstract public function newQuery($excludeDeleted = true);```


Get it finally to create the indexes using the defined $indexSettings

Now I get new errors, related to the (!) related models... ```Argument 1 passed to Textos::isMultiLevelArray() must be of the type array, integer given``` should I keep the conversation in a new issue, or here, as it seems to still be a L4 problem ?

**Comment by timgws** [comment]
I have fixed the _id issue https://github.com/elasticquent/Elasticquent/commit/fbbf6a0e4b5457f89e2e51734154948c022bc6a3

Sounds like `newQuery` is not compatible with Ardent... we will need to change the method name. For the moment, is it possible to revert to 1.0.5, and then we can fix the issue up here and ping you when it's done?

**Comment by MarcosBL** [comment]
For now I have "fixed" the issue locally adding the ```newQuery($excludeDeleted = true)``` parameter to the function definition.

As per the ```Argument 1 passed to Textos::isMultiLevelArray() must be of the type array, integer given``` I fixed it locally by brute force, commenting the whole code inside ```loadRelationsAttributesRecursive()```, the problem seems to be $value receiving... values, not arrays.

I'll wait for the update, thank you very much for your effort, of course

**Comment by timgws** [comment]
@MarcosBL can you please show the query that you are performing to cause this error?

**Comment by MarcosBL** [comment]
Sure, it's a very basic Model (Texts = Title + Content), and I'm just trying to include it in Elasticquent. There is only a couple of relations, with User and Texts itself (a Text have a 1-deep level of Parent-Child with the same table) but I really doubt that could be the problem.

Relevant Model, with vars translated from Spanish:

```
<?php
use LaravelBook\Ardent\Ardent;
use Cviebrock\EloquentSluggable\SluggableInterface;
use Cviebrock\EloquentSluggable\SluggableTrait;
use Elasticquent\ElasticquentTrait;

class Textos extends Ardent implements SluggableInterface {
	use ElasticquentTrait;
    protected $indexSettings = [
        'analysis' => [
            'analyzer' => [
                'folding' => [
                    'tokenizer' => 'standard',
                    'filter' => [ 'standard', 'lowercase', 'asciifolding', 'spanish_stemmer', 'spanish_snowball' ],
                ],
            ],
            'filter' => [
                'spanish_stemmer' => [
                    "type" => "stemmer",
                    "name" => "spanish"
                ],
                'spanish_snowball' => [
                    "type" => "snowball",
                    "language" => "Spanish"
                ]
            ]
        ],
    ];
	protected $mappingProperties = [
		'titulo' => [ 'type' => 'string', 'analyzer' => 'folding' ],
		'texto' => [ 'type' => 'string', 'analyzer' => 'folding' ],
	];
	use SluggableTrait;
	protected $sluggable = array(
		'build_from' => 'titulo',
		'save_to'    => 'slug',
		'on_update'	=> false,
		);
	public $autoHydrateEntityFromInput = true;
	public $forceEntityHydrationFromInput = true;
	public $autoPurgeRedundantAttributes = true;
	public static $rules = array (
		'title' => 'required|between:4,80|unique:textos',
		'content' => 'required|min:50',
		'publisheddate' => 'required|date_format:Y-m-d H:i:s'
		);
	protected $table = 'texts';

	protected $guarded = array('id', '_token', 'submit', 'page', 'image');

	public function afterSave() {
		$this->addToIndex();
	}
	public function beforeCreate() {
		$user = App::make('user');
		$this->created_by = $user->id;
	}
	public function beforeUpdate() {
		$user = App::make('user');
		$this->updated_by = $user->id;
	}
	public function Creator()
	{
		return $this->hasOne('UserProfile', 'user_id', 'created_by');
	}

	public function Editor()
	{
		return $this->hasOne('UserProfile', 'user_id', 'updated_by');
	}

	public function Parent()
	{
		return $this->hasOne('Texts', 'id', 'parent');
	}

	public function Siblings()
	{
		return $this->hasMany('Texts', 'parent', 'parent')->where('parent','<>',0)->where('visible', '=', 1)->orderBy('publisheddate', 'ASC')->orderBy('title', 'ASC');
	}

	public function Child()
	{
		return $this->hasMany('Texts', 'parent', 'id')->where('visible', '=', 1)->orderBy('publisheddate', 'ASC')->orderBy('title', 'ASC');
	}

	public function getContentAttribute() {
		return str_replace(Config::get('local/app.settings.cdn'), Config::get('app.settings.cdn'), $this->attributes['texto']);
	}
}
```

The error shown is 
```Argument 1 passed to Textos::isMultiLevelArray() must be of the type array, integer given```
in the function ```loadRelationsAttributesRecursive```

When it reaches line number 731 ```if (!static::isMultiLevelArray($value)) {``` part, $values contain integers/values in my app, and it seems to expect only Arrays.

If I ```dump($attributes);``` after line 720, I get this:

```
array:11 [▼
  "id" => 5
  "created_at" => "2016-03-23 02:34:06"
  "updated_at" => "2016-03-23 02:57:53"
  "created_by" => 1
  "updated_by" => 1
  "publisheddate" => "2016-04-01"
  "titulo" => "Recursos Humanos"
  "slug" => "recursos-humanos"
  "texto" => "HUGE BUNCH OF TEXT CONTENT"
  "visible" => 1
  "parent" => 1
]
```

I can only wonder the function expects a different input. 

Of course, count on me to do any test you need.

**Comment by timgws** [comment]
Thanks for this @MarcosBL.

I am currently working on another component of my project at work (and my ES stuff is working as it should), so I will not be able to get to this for another 9-10 hours. I hope this is OK with you.

@chuangbo, can you please have a look at this for me?

**Comment by chuangbo** [comment]
Oh I understand now. Because the foreign key has the same name as the relation. And you didn't index the relation into elasticsearch but elasticquent try to load it.

I'll look into it in about 4 hours.

**Comment by MarcosBL** [comment]
I'm not sure but I don't think the problem is the relation name, I translated it for easy of comunication, but the relations are Hermanas(Siblings), Padre(Parent) and Hijas(Child). Creator and Editor are the only ones that are in english in the original.

Of course, @timgws @chuangbo it doesn't matter if it is 10 hours or 10 days! This is an Open Source project, I'm happy to help and grateful for your work!

**Comment by chuangbo** [comment]
@MarcosBL So you have Spanish `Padre` as relation name but English `parent` as foreign key?

**Comment by MarcosBL** [comment]
No, "padre" everywhere, just translated for you to better understand the relations.

Sorry if I confused you even more; I better paste the whole thing "as is" and if you can't understand something, I can translate. The whole file is this:

```
<?php
use LaravelBook\Ardent\Ardent;
use Cviebrock\EloquentSluggable\SluggableInterface;
use Cviebrock\EloquentSluggable\SluggableTrait;
use Elasticquent\ElasticquentTrait;

class Textos extends Ardent implements SluggableInterface {
	use ElasticquentTrait;
    protected $indexSettings = [
        'analysis' => [
            'analyzer' => [
                'folding' => [
                    'tokenizer' => 'standard',
                    'filter' => [ 'standard', 'lowercase', 'asciifolding', 'spanish_stemmer', 'spanish_snowball' ],
                ],
            ],
            'filter' => [
                'spanish_stemmer' => [
                    "type" => "stemmer",
                    "name" => "spanish"
                ],
                'spanish_snowball' => [
                    "type" => "snowball",
                    "language" => "Spanish"
                ]
            ]
        ],
    ];
	protected $mappingProperties = [
		'titulo' => [ 'type' => 'string', 'analyzer' => 'folding' ],
		'texto' => [ 'type' => 'string', 'analyzer' => 'folding' ],
	];
	use SluggableTrait;
	protected $sluggable = array(
		'build_from' => 'titulo',
		'save_to'    => 'slug',
		'on_update'	=> false,
		);
	public $autoHydrateEntityFromInput = true;
	public $forceEntityHydrationFromInput = true;
	public $autoPurgeRedundantAttributes = true;
	public static $rules = array (
		'titulo' => 'required|between:4,80|unique:textos',
		'texto' => 'required|min:50',
		'fecha' => 'required|date_format:Y-m-d H:i:s'
		);
	protected $table = 'textos';

	protected $guarded = array('id', '_token', 'submit', 'page', 'imagen');

	public function afterSave() {
		$this->addToIndex();
	}
	public function beforeCreate() {
		$user = App::make('user');
		$this->created_by = $user->id;
	}
	public function beforeUpdate() {
		$user = App::make('user');
		$this->updated_by = $user->id;
	}
	public function Creator()
	{
		return $this->hasOne('UserProfile', 'user_id', 'created_by');
	}

	public function Editor()
	{
		return $this->hasOne('UserProfile', 'user_id', 'updated_by');
	}

	public function Padre()
	{
		return $this->hasOne('Textos', 'id', 'padre');
	}

	public function Hermanas()
	{
		return $this->hasMany('Textos', 'padre', 'padre')->where('padre','<>',0)->where('esvisible', '=', 1)->orderBy('fecha', 'ASC')->orderBy('titulo', 'ASC');
	}

	public function Hijas()
	{
		return $this->hasMany('Textos', 'padre', 'id')->where('esvisible', '=', 1)->orderBy('fecha', 'ASC')->orderBy('titulo', 'ASC');
	}

	public function getTextoAttribute() {
		return str_replace(Config::get('local/app.settings.cdn'), Config::get('app.settings.cdn'), $this->attributes['texto']);
	}

	public function getFechablogAttribute()
	{
		$date = strtotime($this->attributes['fecha']);
		setlocale(LC_ALL,"es_ES.UTF8","es_ES","esp");
		return (object)[
		"day" => strftime("%d", $date),
		"month" => strftime("%B", $date),
		"year" => strftime("%Y", $date)
		];
	}
	public function getFechaAttribute()
	{
		return date('d/m/Y', strtotime($this->attributes['fecha']));
	}
	public function setFechaAttribute()
	{
		if (trim(Input::get('fecha'))) {
			$newdate = DateTime::createFromFormat('d/m/Y H:i:s', Input::get('fecha')." 00:00:00");
			if ($newdate) {
				$this->attributes["fecha"] = $newdate->format('Y-m-d H:i:s');
			}
			else
			{
				return false;
			}
		}
		else
		{
			$this->attributes["fecha"] = "0000-00-00 00:00:00";
		}
	}
	public function scopeSelect($query, $title = false, $self=false) {
		if ($title) {
			$selectVals[''] = $title;
			$selectVals += $this->where('padre', '=', 0)->where('id', '<>', $self)->orderBy('titulo','ASC')->lists('titulo', 'id');
			return $selectVals;
		}
		else
		{
			return $this->where('padre', '=', 0)->where('id', '<>', $self)->orderBy('titulo','ASC')->lists('titulo', 'id');
		}
	}
}
```

The thing seems to be the ```static::isMultiLevelArray($value)``` expecting always an Array and receiving single values instead.

**Comment by chuangbo** [comment]
@MarcosBL Thank you for your help.

I am still guessing the reason is very likely that the same name between foreign key and relation name. I suggest to use a `padre_id` as foreign name (which also the laravel default behaviour) because:

1. You can not make use of Eloquent relation cache (Eloquent's dynamic properties)
2. If you index the relation `padre` into Elasticquent (eager load), you will lost the padre id, as `$instance->toJson()` (which Elasticquent use to dump document into elasticsearch) use relations first if the relation name conflicts with attribute's name.
3. Actually when I testing this in vanilla laravel 5 without Elasticquent, it doesn't work at all.
```php
class Book extends Model
{
    public function owner()
    {
        return $this->belongsTo('User', 'owner');
    }
}

$book = new Bool();
$book->owner();
PHP error:  Undefined property: App\Book::$owner in /app/vendor/laravel/framework/src/Illuminate/Database/Eloquent/Relations/BelongsTo.php on line 75
```

Can you please try to use the foreign keys with a `_id` postfix and test if possible? Tell me if you get any luck.

If it still failed, could you please put a dump at loadRelationsAttributesRecursive like
```php
                    if ($relation instanceof Relation) {
                        dump($key, $value);
                        // Check if the relation field is single model or collections
                        if (!static::isMultiLevelArray($value)) {
```

This will help us know what attribute cause the error.

**Comment by MarcosBL** [comment]
Of course @chuangbo you were right: to not conflict with other parts of the app, I renamed the relation name (instead of the table field, just for this test, will fix it properly now) so instead of a Padre() relation I just renamed the declaration and calls as RelatedPadre(), prepending "Related" in front of each relation to not conflict with the field name: **BOOM**, working, first try!

So in the end, about the original problem making Elasticquent not compatible it got reduced to just adding a single parameter to ```EloquentMethods.php``` function ```newQuery```

```
    /**
     * Get a new query builder for the model's table.
     */
    abstract public function newQuery($excludeDeleted = true);
```

I'm not that fluent with git workflow, and it's a ridiculous contribution, so I better prefer if you can commit it when you have the time.

Thanks again!

**Comment by chuangbo** [comment]
Great and congrats!

In terms of `newQuery`, I noticed the parameter has been removed at [2 years ago](https://github.com/laravel/framework/commit/8d8b8ce6f0f539f903b8df0f117732451364aed2#diff-d8e24a5815a26f7211e66db787be647fL1620), which included in laravel 4.2.

And actually I don't see what's the reason we have `\Elasticquent\EloquentMethods`. And to remove it from Elasticquent will resolve your problem.

So we could either:
1. [Upgrade to laravel 4.2](https://laravel.com/docs/4.2/upgrade#upgrade-4.2), or
2. Remove `\Elasticquent\EloquentMethods`

What do you think @timgws?

**Comment by MarcosBL** [comment]
I agree with you it doesn't seem to do anything, but I'm not sure if removing ```EloquentMethods``` would break some other compatibility (just starting with Laravel myself). 

Ardent 2.5 (latest L4 compatible version) defines it at [Ardent.php#L925](https://github.com/laravel-ardent/ardent/blob/v2.5.0/src/LaravelBook/Ardent/Ardent.php#L925) with the parameter, so I really don't know if EloquentMethods have any real usage. 

If you plan to just add the parameter to the function, I can confirm that keeps the Ardent compatibility, is what I did for the time being.

If you plan to delete it, I can test if it keeps the Ardent compatibility as well as L4, of course.

**Comment by timgws** [comment]
This was added by @petercoles https://github.com/elasticquent/Elasticquent/commit/116649044b7e92568defca26bf403db3508364ee

**Comment by timgws** [comment]
@chuangbo removing `EloquentMethods` sounds like a good idea at this point, IMHO. It's creating more problems then it solves.

**Comment by petercoles** [comment]
As background, the EloquentMethods trait makes explicit those methods that the ElasticquentTrait relies on from the class being extended, generally Eloquent. As such it doesn't "do" anything, but is a messenger (as in "don't shoot the messenger) highlighting any incompatibilities between what Elasticquent expects and what the class to which it adding delivers. In effect, it's the equivalent of an interface for a class.

@chuangbo is right that the $excludeDeleted parameter was present in Eloquent up to (and including  Laravel (Illuminate/Database) 4.1, but was dropped from 4.2 onwards, which is why this constraint failed in @MarcosBL's project, but worked when it was added back to the method definition. My apologies for not spotting this when I added the constraint.

As an aside, I don't believe that this specific issue is affected by Ardent, which doesn't appear to interfere with Eloquent in this area.

For a quick fix, the best solution would be to remove EloquentMethods from ElasticquentTrait's use statement. this would allow it to still be added optionally by those projects that need this level of reassurance that dependencies are being delivered as expected e.g.
```
class myModel extends Eloquent [or Ardent]
{
    use Elasticquent, EloquentMethods;
    ...
}

Longer term supporting old contexts is a problem that projects such as this tend to run into as they become established and it might be introduce some semantic versioning based on the underlying Laravel versions they support. there are several instances now of slightly hacky code in the package designed to detect the version of Laravel and act differently according to what's found.
